File: proxy.py
# Import FastAPI framework and supporting tools
from fastapi import FastAPI
from pydantic import BaseModel
from fastapi.middleware.cors import CORSMiddleware
from gradio_client import Client
import requests
import os

# 1. Create FastAPI app
app = FastAPI(title="Peptide Function Proxy API")

# 2. Enable CORS
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)

# 3. Initialize HF Space client
# Store space name in Vercel client_key and value in env variables and HF variables and secrets
AMP_SPACE = os.getenv("AMP_SPACE")
client = Client(AMP_SPACE)

# --- Google Sheets logging constants ---
SHEET_URL = os.getenv("SHEET_URL")
SECRET_TOKEN = os.getenv("SECRET_TOKEN")

# 4. Request model
class SequenceRequest(BaseModel):
    sequence: str
    user: str = "anonymous"

# 5. Health check
@app.get("/")
def root():
    return {"message": "Proxy server running"}

# 6. Prediction endpoint
@app.post("/predict")
def predict(req: SequenceRequest):
    try:
        print("Received sequence:", repr(req.sequence))

        # --- HF prediction ---
        result = client.predict(
            sequence=req.sequence,
            api_name="/predict_peptide"
        )
        print("HF raw:", result)

        predictions = []

        if isinstance(result, dict) and "data" in result:
            for row in result["data"]:
                predictions.append({"target": row[0], "probability": float(row[1])})
        elif isinstance(result, list):
            for row in result:
                predictions.append({"target": row[0], "probability": float(row[1])})

        # ⭐ ADDED: Extract ONLY the probabilities for columns D–H
        hf_numbers = [p["probability"] for p in predictions]  # [0.1562, 0.3115, ...]
        print("hf_numbers:", hf_numbers)
        # ⭐ END ADDITION

        # --- POST to Google Sheet ---
        try:
            sheet_response = requests.post(
                url=SHEET_URL,
                headers={"Content-Type": "application/json"},
                json={
                    "sequence": req.sequence,
                    "user": req.user,
                    "source": "iOS app",
                    "token": SECRET_TOKEN,

                    # ⭐ ADDED: send numeric array only
                    "hf_numbers": hf_numbers
                    # ⭐ END ADDITION
                }
            )
            print("Sheet response:", sheet_response.text)
        except Exception as sheet_error:
            print("Failed to log to sheet:", sheet_error)

        return {
            "sequence": req.sequence,
            "predictions": predictions,
            "hf_numbers": hf_numbers  # optional
        }

    except Exception as e:
        print("Error:", e)
        return {
            "sequence": req.sequence,
            "predictions": [],
            "error": str(e)
        }

File: proxy.js
export const config = {
  runtime: "edge"
};

export default async function handler(req) {
  try {
    if (req.method === "GET") {
      return new Response(
        JSON.stringify({ message: "Proxy server running" }),
        { status: 200, headers: { "Content-Type": "application/json" } }
      );
    }

    if (req.method !== "POST") {
      return new Response("Method not allowed", { status: 405 });
    }

    const body = await req.json();
    const sequence = body.sequence || "";

    console.log("Received sequence:", sequence);

    // --- HF SPACE URL (equivalent to your gradio_client call) ---
    const HF_URL =
      "https://Ym420-peptide-function-classification.hf.space/predict_peptide";

    // --- Call HF Space ---
    const hfRes = await fetch(HF_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({ sequence })
    });

    const result = await hfRes.json();
    console.log("HF raw result:", result);

    // --- Normalise prediction output ---
    let predictions = [];

    // Case 1: Gradio DataFrame format
    if (result && result.data) {
      for (const row of result.data) {
        predictions.push({
          target: row[0],
          probability: parseFloat(row[1])
        });
      }
    }

    // Case 2: Simple list of lists
    else if (Array.isArray(result)) {
      for (const row of result) {
        predictions.push({
          target: row[0],
          probability: parseFloat(row[1])
        });
      }
    }

    return new Response(
      JSON.stringify({
        sequence,
        predictions
      }),
      {
        status: 200,
        headers: {
          "Content-Type": "application/json",
          "Access-Control-Allow-Origin": "*"
        }
      }
    );
  } catch (err) {
    console.error("Error in /predict:", err);

    return new Response(
      JSON.stringify({
        sequence: "",
        predictions: [],
        error: String(err)
      }),
      { status: 500, headers: { "Content-Type": "application/json" } }
    );
  }
}
